### PR TITLE
Improve performance of advanced indexing backward

### LINKED ISSUE #20457: index_put_ no longer accepts indices with non-matching backend
## 🐛 Bug
in index_put_ operation indices backend now has to match source backend. It used to be not necessary, so that e.g. cuda tensor could be indexed by cpu tensor. Blame points to #17991  https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Indexing.cpp#L204

## To Reproduce
```
import torch

a=torch.arange(30, dtype=torch.float).view(5,6).cuda()
ind0 = torch.arange(0,a.size(0), step=2)
gO = torch.randn(a[ind0].size()).cuda()
a.index_put_((ind0,), gO, accumulate=True)
torch.cuda.synchronize()
```
used to work, now it does not. Note that forward operation with ind0 on the cpu still works (`a[ind0]`), but index_put_ breaks
## Expected behavior
Either clarification in the docs and in the code that indices tensor has to have the same backend as source tensor, or fixing the RuntimeError. In case we want to disable indexing with a different backend, it would make sense to disable it for the forward indexing operation too, right now forward would be fine, but backward will through a runtime error. 

## Environment
Pytorch 1.1 binary and recent source builds

cc @colesbury

### PR DESCRIPTION
**Pull request by ngimel**
This PR improves performance of advanced indexing backward, partially solving #15245 (performance is still worse than gather, but not by such outrageous margins). Before, using benchmarking harness from #15245, cuda 10/V100:
```
Indexing is faster by at most -270.61607820767887 us on N: 16 D: 256 K: 1
Indexing is slower by at most 11127.466280784833 us on N: 16 D: 4096 K: 4096
```
after:
```
Indexing is faster by at most 23.524456737696028 us on N: 512 D: 4096 K: 4096
Indexing is slower by at most 186.24056029472553 us on N: 16 D: 1024 K: 4096
```
Strategy is to reuse embedding backward kernel, adapting it to handle unindexed dimensions in the beginning by launching additional threadblocks, and also allowing it to handle slices that are bigger than `65K*128`, that is hardly ever a problem for embedding. Still, integer indexing is baked in the kernel, and is important for performance, so for now bigger than 2G element tensors are not supported. 
The main savings come from not having to expand index to all unindexed dimensions, and not sorting expanded index with incoming gradient values, but rather only sorting unexpanded index. 
There are ways to make sorting overhead smaller (thanks @mcarilli for suggestions) but I'll get to it when it becomes a real problem, or rather, when cuda graphs will force us to get rid of thrust::sort calls. 
I've also added tests for indexing backward, before tests for index_put_ and indexing backward were non-existent. 
This PR also fixes #20457 by casting indices to `self` backend.

### CODE REVIEW

**Comment by ngimel** [comment]
CI today is failing with 
```
May 15 22:32:24 CMake Error at third_party/fbgemm/CMakeLists.txt:115 (add_subdirectory):
May 15 22:32:24   The source directory
May 15 22:32:24 
May 15 22:32:24     /var/lib/jenkins/workspace/third_party/fbgemm/third_party/asmjit
May 15 22:32:24 
May 15 22:32:24   does not contain a CMakeLists.txt file.
May 15 22:32:24 
```
(also in another PR), looks like fbgemm submodule is not recursively updated. Can I do something on my side? I've just merged master, so I don't think rebase would help?

**Comment by ezyang** [comment]
@pytorchbot rebase this please

should be fixed on master

**Comment by ezyang** [comment]
I'll let @colesbury take a first whack at this.

**Comment by ngimel** [comment]
I've pushed what I hope is a fix for rocm failures (and removed tabs), but I think same failures should be happening in embedding backward, because in some places WARP_SIZE is assumed to be 32, in some places a device-dependent #define'd WARP_SIZE is used. Is embedding backward not compiled/tested on rocm?

**Comment by ezyang** [comment]
cc @iotamudelta @bddppq re Natalia's question

**Comment by ezyang** [comment]
@colesbury Let me know if you want me to attempt a review (the PR seems pretty involved, heh heh)

**Comment by iotamudelta** [comment]
The fix for ROCm to change from assuming WARP_SIZE 32 to 64 makes sense. Embedding is certainly compiled in ROCm (it's not in the skipped functins), cannot comment on tests off the top of my head. Thanks!

**Comment by iotamudelta** [comment]
@ezyang maybe we should aim to have one warp size defined somewhere based on whatever architecture we compile for and that being reused everywhere as opposed to the wild ifdef'ing and hardcoding that has happened so far? What header would be a good place for something like this to go?

**Comment by ezyang** [comment]
> we should aim to have one warp size defined somewhere based on whatever architecture we compile for and that being reused everywhere as opposed to the wild ifdef'ing and hardcoding that has happened so far? What header would be a good place for something like this to go?

Yes, this sounds reasonable. Maybe something like c10/macros/Macros.h  or c10/cuda/CUDAMacros.h

**Comment by iotamudelta** [comment]
For obvious reasons I prefer c10/macros/Macros.h - I can get started on that.

**Comment by colesbury** [comment]
I skimmed the code and it looks reasonable, but I don't have time to do a real review for the next few days. @ezyang if you want to review it (or get someone else to review it), that would be great

**Review by soumith — APPROVED** [review]
Sam says he reviewed it and it looks good, but he didn't approve it.

**Comment by ngimel** [comment]
The problem I currently have is that I removed boundary checking for indices, because it is a big performance overhead, and by the time of backward, if indices were invalid, forward has errored out, so they are not necessary. However, `index_put` is a user facing function, and a user may directly call it passing invalid indices, so unconditionally removing index checks is not right. Can we have, e.g., `index_put_unsafe` for index gradient without checks, and regular index_put that does index checks? We already have similar thing with _sparse_coo_tensor_unsafe that is used in a few backward functions.

**Comment by soumith** [comment]
@ngimel `index_put_unsafe` seems reasonable, or even better `index_put(unsafe=True)`

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: FYI, you don't need to construct the stringstream manuallyl; `AT_INDEX_ERROR` takes any number of comma arguments and will cat them together using a stringstream. Also, this macro got a new name; it's `TORCH_CHECK_INDEX(false, ...)` now (No need to fix either of these issues unless you happen to be addressing other fixes on this PR).

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: cc @izdeby (I don't think there's anything wrong with this code, but I saw the instance of `kBool` and thought you'd might like to know about it)

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: nit: `const auto&`

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: This comment is slightly confusing: you refer to `masks` but there is no argument named masks here. Did you mean `indices`?

Comment (ezyang) [reply aten/src/ATen/native/IndexingUtils.h]: Reading the code, I guess what you mean is, if an element of indices is an index tensor, then just return it directly; otherwise convert if from boolean mask to indices mask.

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [reply aten/src/ATen/native/IndexingUtils.h]: You will probably want to axe the `kByte` case eventually (I don't know what order you are doing these things.)

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: You never seem to set `special_empty` to true.

**Comment by ngimel** [comment]
@ezyang all your criticisms are valid, however, this is just copy paste of functions from Indexing.cpp to IndexingUtils.h, so I tried to keep changes to a minimum :-)

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: nit: It is a little confusing to use the word "contiguous" here since normally it refers to contiguity of strides (no action needed, just remarking)

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h]: Nice comment! Thank you!

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: s/thei/this/

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: has the same what?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: nit: processed

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: These should all be `TORCH_CHECK`, because a user in Python can write code that triggers these errors. (Remember: `TORCH_INTERNAL_ASSERT` is for any checks that would be OK to delete, because they *really really* should never be violated.)

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: I guess, the indexes can have the same value when our slice size is greater than 1?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/Indexing.cu]: I find the use of variable name `input` here kind of confusing, because, based on the call site, this isn't actually an input in that sense; it's `sorted_indices`, no?

File: aten/src/ATen/native/cuda/Indexing.cu
Comment (ezyang) [reply aten/src/ATen/native/cuda/Indexing.cu]: Oh, is this related to the removal of indexing checks you mentioned?

**Review by ezyang — DISMISSED** [review]
I did a closer but still somewhat superficial review. Some comments here and there, nothing major, happy to merge and defer comments for later if necessary.

**Comment by colesbury** [comment]
@soumith @ngimel I'd like to avoid expanding the public API, so I'd prefer something like: `_index_put_unsafe` (i.e. prefixed with an underscore) as opposed to a kwarg.

@ngimel what is the expensive bounds checks? Is it `wrapIndexOnce`? I assume actual `assert` in device code is relatively fast.

**Comment by ngimel** [comment]
Yes, it's wrapIndexOnce, that runs a few (on the order of 10) launch latency bound kernels and adding 20-30 us. I've just done kwarg, but I'll redo as separate API. asserts in the device code in forward are not a problem.

**Comment by ngimel** [comment]
Added non-user-facing unsafe index_put option, addressed @ezyang's comments. I still plead copy paste.

**Comment by ezyang** [comment]
Waiting on CI to land

**Pull request closed without merging**
